### ISSUE TITLE
torch.optim

### ISSUE THREAD

**Issue by Ludobico** [Ludobico/Python_AL_01#32]
# torch.optim
`torch.optim`은 Pytorch에서 제공하는 최적화 알고리즘을 사용하기 위한 패키지입니다. 일반적으로 신경망을 학습시키기 위해서는 손실 함수를 최소화하는 가중치를 찾아야 합니다. 이를 위해서 우리는 최적화 알고리즘을 사용합니다.
torch.optim 패키지는 이러한 알고리즘을 간단하게 사용할 수 있게 해줍니다.

torch.optim 패키지에서 제공하는 최적화 알고리즘 종류는 다음과 같습니다.

* SGD (Stochastic Gradient Descent)
* Adam
* Adagrad
* Adadelta
* Adamax
* RMSprop

# torch.optim.zero_grad
`torch.optim.zero_grad()`는 가중치 텐서에 저장된 기울기 정보를 초기화하는 함수입니다. 이 함수는 주로 역전파 이후에 호출됩니다.

기울기 정보는 가중치 텐서의 grad 속성에 저장되고 있습니다. 이 속성은 기울기를 저장할 수 있는 텐서입니다.
이 속성은 `requires_grad=True` 속성을 가진 가중치 텐서에만 정의됩니다.

zero_grad() 함수는 기울기 정보를 초기화하기 위해 grad 텐서의 모든 요소를 0으로 설정합니다. 이렇게 하면 이전 역전파 정보가 삭제되고, 새로운 역전파 정보를 저장할 수 있게됩니다.

# torch.optim.step
`torch.optim.step`은 가중치 텐서를 업데이트하기 위한 함수입니다. 이 함수는 최적화기(optimizer)가 정의한 알고리즘을 이용해 가중치 텐서의 값을 업데이트합니다.

업데이트 알고리즘은 손실 함수의 기울기 정보를 이용해 가중치 텐서의 값을 조정합니다. 이렇게 가중치 텐서를 업데이트함으로써 손실 함수의 값을 최소화할 수 있게 됩니다.

# torch.optim.SGD
SGD는 가장 기본적인 최적화 알고리즘입니다. 이 알고리즘은 기울기 업데이트를 수행할 때 전체 데이터 셋을 사용하는 것이 아니라, 일부 샘플(batch)을 추출해 기울기를 게산합니다. 이렇게 계산된 기울기를 이용해 손실 함수의 값을 최소화하는 가중치를 찾습니다.

torch.optim 패키지에서 SGD를 사용하기 위해서는 `torch.optim.SGD` 클래스를 생성해야 합니다. 생성자에 가중치 텐서들을 전달하고, `step` 메서드를 호출하면 가중치가 업데이트됩니다.

다음은 SGD 예제 코드입니다.
```py
import torch

# 가중치 텐서를 정의합니다.
weights = torch.tensor([1.0, 2.0], requires_grad=True)

# SGD 최적화기를 정의합니다. 이 예제에서는 학습률을 0.1로 설정합니다.
optimizer = torch.optim.SGD([weights], lr=0.1)

# 손실 함수를 정의합니다.
def loss_fn(weights):
  return (weights**2).sum()

# 최적화를 100번 수행합니다.
for _ in range(100):
  # 손실 함수를 계산하고 역전파합니다.
  loss = loss_fn(weights)
  loss.backward()

  # 가중치를 업데이트합니다.
  optimizer.step()

  # 역전파 정보와 가중치의 grad를 초기화합니다.
  optimizer.zero_grad()

# 최종 가중치
print(weights)
```
